### PR TITLE
[SPARK-14000][SQL] case class with a tuple field can't work in Dataset

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

When we validate an encoder, we may call `dataType` on unresolved expressions. This PR fix the validation so that we will resolve attributes first.
## How was this patch tested?

a new test in `DatasetSuite`

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rxin, this should fixes the failure in your https://github.com/apache/spark/pull/11737

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 1]
```diff
@@ -110,7 +110,12 @@ case class GetStructField(child: Expression, ordinal: Int, name: Option[String]
 
   override def dataType: DataType = childSchema(ordinal).dataType
   override def nullable: Boolean = child.nullable || childSchema(ordinal).nullable
-  override def toString: String = s"$child.${name.getOrElse(childSchema(ordinal).name)}"
+
+  override def toString: String = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: kind of unrelated change, `toString` should be able to run even this expression is unresolved.

**Comment by rxin** [comment]
thanks - looks good to me (but i don't know this code well enough).

**Comment by cloud-fan** [comment]
cc @marmbrus

**Comment by liancheng** [comment]
LGTM, merging to master.
